### PR TITLE
Name previously unnamed thread pool threads

### PR DESCRIPTION
**Pull request by steviez**
#### Problem
See https://github.com/anza-xyz/agave/issues/34 for context

#### Summary of Changes
- Name the threads in thread pools that were previously unnamed

A few notes on the tokio pools:
- Both pools look to be heavily over-provisioned; however, that is not the goal of this PR and will be addressed in further work
- It seemed better to import tokio directly over using the re-export from `jsonrpc_server_utils`
- `Runtime::new()` calls `Builder::new_multi_thread().enable_all()` under the hood, so there is no change in behavior here; builder allows us to name the threads
https://github.com/tokio-rs/tokio/blob/1f924f95f12c362f5116c560acfa005149f110a1/tokio/src/runtime/runtime.rs#L181-L183

### CODE REVIEW

File: net-utils/src/ip_echo_server.rs [hunk 1]
```diff
@@ -173,7 +173,11 @@ pub fn ip_echo_server(
 ) -> IpEchoServer {
     tcp_listener.set_nonblocking(true).unwrap();
 
-    let runtime = Runtime::new().expect("Failed to create Runtime");
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .thread_name("solIpEchoSrvrRt")
+        .enable_all()
+        .build()
+        .unwrap();
```
Comment (bw-solana) [inline net-utils/src/ip_echo_server.rs]: Keep the expect instead?

File: runtime/src/snapshot_utils/snapshot_storage_rebuilder.rs [hunk 2]
```diff
@@ -418,6 +418,7 @@ impl SnapshotStorageRebuilder {
     /// Builds thread pool to rebuild with
     fn build_thread_pool(&self) -> ThreadPool {
         ThreadPoolBuilder::default()
+            .thread_name(|i| format!("solRbuildSnap{i:02}"))
```
Comment (bw-solana) [inline runtime/src/snapshot_utils/snapshot_storage_rebuilder.rs:421]: If we squeeze the threadpools down, we'll be able to get that e into Rebuild

File: validator/src/admin_rpc_service.rs [hunk 3]
```diff
@@ -815,8 +815,12 @@ pub async fn connect(ledger_path: &Path) -> std::result::Result<gen_client::Clie
...
 }
 
-pub fn runtime() -> jsonrpc_server_utils::tokio::runtime::Runtime {
-    jsonrpc_server_utils::tokio::runtime::Runtime::new().expect("new tokio runtime")
+pub fn runtime() -> Runtime {
+    tokio::runtime::Builder::new_multi_thread()
+        .thread_name("solAdminRpcRt")
+        .enable_all()
+        .build()
+        .unwrap()
```
Comment (bw-solana) [inline validator/src/admin_rpc_service.rs]: Keep the expect?

File: net-utils/src/ip_echo_server.rs [hunk 4]
```diff
@@ -173,7 +173,11 @@ pub fn ip_echo_server(
 ) -> IpEchoServer {
     tcp_listener.set_nonblocking(true).unwrap();
 
-    let runtime = Runtime::new().expect("Failed to create Runtime");
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .thread_name("solIpEchoSrvrRt")
+        .enable_all()
```
Comment (bw-solana) [inline net-utils/src/ip_echo_server.rs:178]: I would imagine this enables more features than the default configuration. Do we know what that delta looks like and if there's any meaningful performance/resource difference?

**Review by bw-solana — COMMENTED** [review]
Looks reasonable to me. My only concern is around which tokio features are now getting enabled (probably just my own ignorance)

File: net-utils/src/ip_echo_server.rs (hunk 4 above)
Comment (steviez) [reply net-utils/src/ip_echo_server.rs:178]: Nope, no difference. I included a note about this in the PR description, but marking up the diff would have been more visible. In any case, `Runtime::new()` calls `Builder::new_multi_thread().enable_all()` under the hood. So, these are equivalent:
https://github.com/tokio-rs/tokio/blob/1f924f95f12c362f5116c560acfa005149f110a1/tokio/src/runtime/runtime.rs#L181-L183

File: runtime/src/snapshot_utils/snapshot_storage_rebuilder.rs (hunk 2 above)
Comment (steviez) [reply runtime/src/snapshot_utils/snapshot_storage_rebuilder.rs:421]: Potentially yeah.

These thread are only around during startup, and they are used to rebuild an accountsDb. The system doesn't have much else going on at the time, so we might actually be ok with this pool using so many threads (of course assuming that the work it is doing is more efficient with the higher number of parallelism)

File: net-utils/src/ip_echo_server.rs (hunk 1 above)
Comment (steviez) [reply net-utils/src/ip_echo_server.rs]: I'm not a huge fan of those expects. To me, is is comparable to how we just `.unwrap()` every time we try to acquire a `RwLock` or `Mutex`.

Even more proper would be bubbling an error up and halting the validator; this would be simple enough to do and I can tackle that. But, I would rather do that separate from this PR

File: validator/src/admin_rpc_service.rs (hunk 3 above)
Comment (steviez) [reply validator/src/admin_rpc_service.rs]: Similar note as other comment about the `.unwrap()` vs. `.expect()` here; I don't think the `.expect()` does us much (the backtrace will be more informative than the message) and a better result would be passing an error and letting caller handle the error properly.

Looking at the callers of this `runtime()` function, it looks like there would be a bit more of a refactor to properly handle error. Maybe I can do it while I'm collecting data for thread pool size reductions later on

File: net-utils/src/ip_echo_server.rs (hunk 1 above)
Comment (t-nelson) [reply net-utils/src/ip_echo_server.rs]: the value of `expect()` over `unwrap()` is justifying the reason that you believe it is safe to panic on failure. we should _always_ prefer `expect()` to `unwrap()`. no undefended panicking

File: validator/src/admin_rpc_service.rs [hunk 5]
```diff
@@ -6,7 +6,6 @@ use {
     jsonrpc_ipc_server::{
         tokio::sync::oneshot::channel as oneshot_channel, RequestContext, ServerBuilder,
     },
-    jsonrpc_server_utils::tokio,
```
Comment (t-nelson) [inline validator/src/admin_rpc_service.rs]: can we make this change in a different pr? it's outside the scope of this one

Comment (steviez) [reply validator/src/admin_rpc_service.rs]: https://github.com/anza-xyz/agave/pull/116

File: net-utils/src/ip_echo_server.rs (hunk 4 above)
Comment (bw-solana) [reply net-utils/src/ip_echo_server.rs:178]: I actually read that in the description and somehow discarded the information by the time I got to this code. Thanks!

File: net-utils/src/ip_echo_server.rs (hunk 1 above)
Comment (steviez) [reply net-utils/src/ip_echo_server.rs]: > the value of expect() over unwrap() is justifying the reason that you believe it is safe to panic on failure

I don't think the pre-existing message of `Failed to create Runtime` conveys your sentiment much (if any) better than `.unwrap()`. In this case, we're trying to allocate a resource and to me, the `.unwrap()` implicitly says the same thing of "bail out if we can't get this"

To be clear, I do the see the value in other cases. For example, here is a little snippet from Blockstore. The below `.expect()` message is more informative than a regular `.unwrap()`:
```rust
            .checked_add(1)
            .expect("overflow from trusted value");
```

In any case, I restored the `expect()`'s that I changed to `unwrap()`, and did two bonus `unwrap() ==> expect()` conversions for the Rayon pools in this PR. In that sense, I have not made anything "worse", and I think any additional cleanup can be considered separately.

File: runtime/src/bank.rs [hunk 6]
```diff
@@ -1483,7 +1483,10 @@ impl Bank {
         let epoch = self.epoch();
         let slot = self.slot();
         let (thread_pool, thread_pool_time) = measure!(
-            ThreadPoolBuilder::new().build().unwrap(),
+            ThreadPoolBuilder::new()
```
Comment (t-nelson) [inline runtime/src/bank.rs:1486]: oh gross. i forgot about this one :joy:

**Review by bw-solana — APPROVED** [review]
LGTM

**Review by t-nelson — APPROVED** [review]
man github needs :hide-the-pain-harold:
